driver/mloCumulativeReductionHost.hpp

```diff
@@ -0,0 +1,151 @@
+{
+    a = b;
+    update(c..., d...);
+}
+
+template <typename T, typename... Ts>
+struct reduce_func_base
+{
+    reduce_func_base(){};
+    virtual ~reduce_func_base(){};
+    virtual inline bool isbetter(const T& /*a*/, const T& /*b*/) const { return false; }
+    virtual inline void combine(T& a, T b) const { a = b; }
+    inline constexpr void calculate(T& a, T b, Ts&... c, Ts... d) const
+    {
+        if(!isbetter(a, b))
+        {
+            combine(a, b);
+            update(c..., d...);
+        }
+    }
+};
+
+template <miopenCumOp_t OP, typename T, typename... Ts>
+struct reduce_func : reduce_func_base<T, Ts...>
+{
+    virtual ~reduce_func(){};
+};
+
+template <typename T, typename... Ts>
+struct reduce_func<MIOPEN_CUM_MAX, T, Ts...> : reduce_func_base<T, Ts...>
+{
+    const float START_VAL = -std::numeric_limits<float>::max();
+    inline bool isbetter(const T& a, const T& b) const { return a > b; }
+};
+
+template <typename T, typename... Ts>
+struct reduce_func<MIOPEN_CUM_MIN, T, Ts...> : reduce_func_base<T, Ts...>
+{
+    const float START_VAL = std::numeric_limits<float>::max();
+    inline bool isbetter(const T& a, const T& b) const { return a < b; }
+};
+
+template <typename T, typename... Ts>
+struct reduce_func<MIOPEN_CUM_SUM, T, Ts...> : reduce_func_base<T, Ts...>
+{
+    const float START_VAL = 0.0f;
+    inline void combine(T& a, T b) const { a += b; }
+};
+
+template <typename T, typename... Ts>
+struct reduce_func<MIOPEN_CUM_PROD, T, Ts...> : reduce_func_base<T, Ts...>
+{
+    const float START_VAL = 1.0f;
+    inline void combine(T& a, T b) const { a *= b; }
+};
+
+template <miopenCumOp_t OP, typename Tgpu, typename Tcheck>
+int32_t mloCumulativeReductionForwardRunHost(const miopenTensorDescriptor_t inputDesc,
+                                             const miopenTensorDescriptor_t outputDesc,
+                                             const miopenTensorDescriptor_t indicesDesc,
+                                             const Tgpu* input,
+                                             Tcheck* output_host,
+                                             int* indices_host,
+                                             const int dim,
+                                             const bool exclusive,
+                                             const bool reverse)
+{
+    const int ndims     = miopen::deref(inputDesc).GetNumDims();
+    const auto true_dim = ((dim % ndims) + ndims) % ndims;
+
+    auto input_tv   = miopen::get_inner_expanded_tv<5>(miopen::deref(inputDesc));
+    auto output_tv  = miopen::get_inner_expanded_tv<5>(miopen::deref(outputDesc));
+    auto indices_tv = miopen::get_inner_expanded_tv<5>(miopen::deref(indicesDesc));
+
+    auto size       = miopen::deref(inputDesc).GetElementSize();
+    auto inner_size = miopen::deref(inputDesc).GetLengths()[true_dim];
+    auto outer_size = size / inner_size;
+
+    auto op_worker = reduce_func<OP, float, int>{};
```

Comment:
It's a bad idea using the same function to verify itself.

Reply:
I don't understand your comment clearly. `reduce_func` template struct has been implemented independent with that is used in `MIOpenCumulativeReduction`. In fact, their code are different.